src/account/operations/retry.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use iota_client::{
+    api_types::dto::LedgerInclusionStateDto,
+    block::{
+        payload::{transaction::TransactionId, Payload},
+        Block, BlockId,
+    },
+};
+
+use crate::account::{handle::AccountHandle, types::InclusionState};
+
+const DEFAULT_RETRY_UNTIL_INCLUDED_INTERVAL: u64 = 1;
+const DEFAULT_RETRY_UNTIL_INCLUDED_MAX_AMOUNT: u64 = 40;
+
+impl AccountHandle {
+    /// Retries (promotes or reattaches) a block for provided block id until it's included (referenced by a
+    /// milestone). This function is re-exported from the client library and default interval is as defined in iota.rs.
+    /// Returns the included block at first position and additional reattached blocks
+    pub async fn retry_until_included(
+        &self,
+        block_id: &BlockId,
+        interval: Option<u64>,
+        max_attempts: Option<u64>,
+    ) -> crate::Result<Vec<(BlockId, Block)>> {
+        Ok(self
+            .client
+            .retry_until_included(block_id, interval, max_attempts)
+            .await?)
+    }
+
+    /// Retries (promotes or reattaches) a transaction sent from the account for a provided transaction id until it's
+    /// included (referenced by a milestone). Returns the included block id.
+    pub async fn retry_transaction_until_included(
+        &self,
+        transaction_id: &TransactionId,
+        interval: Option<u64>,
+        max_attempts: Option<u64>,
+    ) -> crate::Result<BlockId> {
+        log::debug!("[retry_transaction_until_included]");
+
+        if let Some(transaction) = self.read().await.transactions.get(transaction_id) {
```

Comment:
The lock is kept the whole duration of the if let scope, which can be very long in the context of this method. If you declare transaction just before, the lock won't be kept.

Reply:
If it complains, I think it's worth cloning the transaction 🤔